src/main.cpp

```diff
@@ -64,43 +94,56 @@ int Main(vector<string> args)
 			else{funcName = "";}
 		}
 
-		if(funcName != ""){
+		if(funcName != ""){ // use the specified function or the default function from the lua script
 			parser[rule.c_str()] = [&L, rule, funcName](const SemanticValues& sv, any& dt){
 
 				// find function				
 				lua_getglobal(L, funcName.c_str());
 
 				// push input parameters on stack
-				lua_pushstring(L, rule.c_str()); // name of the rule
-				lua_pushlstring(L, sv.c_str(), sv.length()); // matched string
-				lua_pushinteger(L, sv.line_info().first); // line
-				lua_pushinteger(L, sv.line_info().second); // column
-				lua_pushinteger(L, sv.choice()); // choice
-				lua_newtable(L); // semantic values
-				for(auto& val:sv){
-					lua_pushstring(L, val.get<LuaReturn>().ruleName.c_str()); // key for table indexing
-					lua_pushvalue(L, val.get<LuaReturn>().stackIndex); // stack pointer to actual content
-					lua_settable(L, -3); // add field to table
-				}
-				lua_newtable(L); // tokens
-				for(int i=0; i<sv.tokens.size(); i++){
-					lua_pushinteger(L, i); // key for table indexing
-					lua_pushlstring(L, sv.tokens[i].first, sv.tokens[i].second); // token
-					lua_settable(L, -3); // add field to table
-				}
-
-				// call function
-				if (lua_pcall(L, 7, 1, 0)){
+				lua_newtable(L);
+					lua_pushfield(L, "rule", rule.c_str());
+					lua_pushfield(L, "matched", sv.c_str(), sv.length());
+					lua_pushfield(L, "line", sv.line_info().first);
+					lua_pushfield(L, "column", sv.line_info().second);
+					lua_pushfield(L, "choice", sv.choice());
+					lua_opensubtable(L, "subnodes"); 
+						for(auto& val:sv){ lua_pushfield(L, val.get<LuaStackPtr>()); }
```

Comment:
This isn't really helpful: for a rule like `Add <- (Mul AddOp)* Mul` it passes a table with two elements `Mul` and `AddOp` to the callback. But the callback requires the entire sequence of matched nodes, so an array should be passed instead.